webdictionary/src/main/java/com/eis0/kademlianetwork/IdFinderHandler.java

```diff
@@ -1,2 +1,158 @@
+                break;
+            case RemoveFromDictionary:
+                findId = RequestTypes.FindIdForDeleteRequest;
+                taskResult = RequestTypes.ResultDeleteRequest;
+            case Refresh:
+                findId = RequestTypes.FindId;
+                taskResult = RequestTypes.SearchResultReplacement;
+        }
+
+        //Obtain the KademliaId belonging to the local node (myself)
+        KademliaId netId = KademliaNetwork.getInstance().getLocalNode().getId();
+        //Obtain the local RoutingTable
+        SMSKademliaRoutingTable table = KademliaNetwork.getInstance().getLocalRoutingTable();
+        //Get the node with the node ID closest to the idToFind
+        SMSKademliaNode closestNode = table.findClosest(idToFind, 1).get(0);
+        SMSKademliaNode nodeToFind = new SMSKademliaNode(idToFind);
+
+        BigInteger idToFindDistanceFromNetId = idToFind.getXorDistance(netId);
+        BigInteger idToFindDistanceFromClosest = idToFind.getXorDistance(closestNode.getId());
+
+        //1. Checking if I'm the searched id (This state should be impossible, it's a fail safe)
+        //2. Checking if inside my RoutingTable there is a node with the ID to find
+        //3. I got further away from what I'm looking for, so I'm the closest one: I return this ID
+        if (netId == idToFind ||
+                KademliaNetwork.getInstance().isNodeInNetwork(nodeToFind) ||
+                idToFindDistanceFromClosest.compareTo(idToFindDistanceFromNetId) > 0) {
+            sendResult(taskResult, idToFind, searcher);
+            retryIfDead(idToFind, searcher, researchMode, searcher);
+            return;
+        }
+        //else
+        //I ask to the closest node inside my Routing Table to continue the research
+        SMSPeer closer = closestNode.getPeer();
+        keepLooking(findId, idToFind, searcher, closer);
+        retryIfDead(idToFind, searcher, researchMode, closer);
+    }
+
+
+    /**
+     * This method sends to the specified target {@link SMSPeer} the result of the previously
+     * carried out research
+     *
+     * @param requestType The type of the result, sent to allow the receiver of the
+     *                    result to define which request the result itself belongs to
+     *                    (There may be multiple pending requests, each of a different nature)
+     * @param idToFind    The ID whose research originated the request
+     *                    It's sent back as response to the research
+     * @param targetPeer  The Peer representing the target that will receive the result
+     */
+    private static void sendResult(RequestTypes requestType, KademliaId idToFind, SMSPeer targetPeer) {
+        KademliaMessage kadMessage = new KademliaMessage(requestType, idToFind, null, null, null);
+        SMSMessage searchResult = new SMSMessage(targetPeer, kadMessage.toString());
+        SMSManager.getInstance().sendMessage(searchResult);
```

Comment:
@turcato-niccolo This is an example of how is created and sent a message with the new structure.
The major add I did was the class KademliaMessage

Reply:
Nice, much better👍